pkg/controllers/metrics/pod/controller.go

```diff
@@ -0,0 +1,186 @@
+
+package pod
+
+import (
+	"context"
+	"fmt"
+	"strings"
+
+	"knative.dev/pkg/logging"
+
+	"github.com/aws/karpenter/pkg/apis/provisioning/v1alpha5"
+	"github.com/prometheus/client_golang/prometheus"
+	v1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/types"
+	controllerruntime "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/manager"
+	crmetrics "sigs.k8s.io/controller-runtime/pkg/metrics"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+)
+
+const (
+	podName             = "name"
+	podNameSpace        = "namespace"
+	ownerSelfLink       = "owner"
+	podHostName         = "node"
+	podProvisioner      = "provisioner"
+	podHostZone         = "zone"
+	podHostArchitecture = "arch"
+	podHostCapacityType = "capacity_type"
+	podHostInstanceType = "instance_type"
+	podPhase            = "phase"
+)
+
+var (
+	podGaugeVec = prometheus.NewGaugeVec(
+		prometheus.GaugeOpts{
+			Namespace: "karpenter",
+			Subsystem: "pods",
+			Name:      "state",
+			Help:      "Pod state.",
+		},
+		labelNames(),
+	)
+)
+
+// Controller for the resource
+type Controller struct {
+	KubeClient client.Client
+	LabelsMap  map[types.NamespacedName]prometheus.Labels
+}
+
+func init() {
+	crmetrics.Registry.MustRegister(podGaugeVec)
+}
+
+func labelNames() []string {
+	return []string{
+		podName,
+		podNameSpace,
+		ownerSelfLink,
+		podHostName,
+		podProvisioner,
+		podHostZone,
+		podHostArchitecture,
+		podHostCapacityType,
+		podHostInstanceType,
+		podPhase,
+	}
+
+}
+
+// NewController constructs a controller instance
+func NewController(kubeClient client.Client) *Controller {
+	return &Controller{
+		KubeClient: kubeClient,
+		LabelsMap:  make(map[types.NamespacedName]prometheus.Labels),
+	}
+}
+
+// Reconcile executes a termination control loop for the resource
+func (c *Controller) Reconcile(ctx context.Context, req reconcile.Request) (reconcile.Result, error) {
+	ctx = logging.WithLogger(ctx, logging.FromContext(ctx).Named("podmetrics").With("pod", req.Name))
+	// Retrieve pod from reconcile request
+	pod := &v1.Pod{}
+	if err := c.KubeClient.Get(ctx, req.NamespacedName, pod); err != nil {
+		if errors.IsNotFound(err) {
+			// Remove gauge due to pod deletion
+			if labels, ok := c.LabelsMap[req.NamespacedName]; ok {
+				podGaugeVec.Delete(labels)
+			} else {
+				logging.FromContext(ctx).Errorf("Failed to delete gauge, failed to locate labels")
+			}
+			return reconcile.Result{}, nil
+		}
+		return reconcile.Result{}, err
+	}
+	// Remove the previous gauge after pod labels are updated
+	if labels, ok := c.LabelsMap[req.NamespacedName]; ok {
+		podGaugeVec.Delete(labels)
+	}
+	if err := c.updateGauges(ctx, pod); err != nil {
+		logging.FromContext(ctx).Errorf("Failed to record metrics, %s", err.Error())
+		return reconcile.Result{}, err
+	}
+	return reconcile.Result{}, nil
+}
+
+func (c *Controller) updateGauges(ctx context.Context, pod *v1.Pod) error {
```

Comment:
Feels odd to me that the two controllers follow different patterns. This controller relies on a helper to generate labels, while the node controller does it all inline. Is there any reason for this? 

Some of these comments may seem like nit picks, and they are, but they are valuable for project health and for the development of your own style. Do your best to be intentional about every line you write. It may feel like it slows you down in the short term, but it will accelerate you in the long term. Happy to discuss more of these ideas in a less public setting.

Reply:
The node controller also calls the generateLabels helper function. But instead of calling it in updateGauges, it calls the helper function in insertGaugeValues. The updateGauges function in the node controller is too long and break the gocyclo test without breaking it down into subroutine (i.e., insertGaugeValues).

I totally agree with you. These suggestions make great sense and can only be learned through experience and careful thinking. I really appreciate them!